daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala

```diff
@@ -400,114 +427,283 @@ abstract class InfosetInputter
+// In Daffodil, the two events will always be back-to-back.
 //
-sealed trait InfosetEventKind
-case object StartKind extends InfosetEventKind { override def toString = "start" }
-case object EndKind extends InfosetEventKind { override def toString = "end" }
+sealed trait InfosetEventKind {
+
+  /**
+   * Converts name like StartArray into "array start"
+   *
+   * Error messages depend on this word order.
+   */
+  override def toString() = {
+    val cn = Misc.getNameFromClass(this)
+    val initialLC = Misc.initialLowerCase(cn)
+    val (se, ea) = initialLC.span(_.isLower)
+    val res = ea.toLowerCase() + " " + se
+    res
+  }
+}
+
+/**
+ * InfosetEvents are of these kinds.
+ */
+object InfosetEventKind {
+  sealed trait StartKind extends InfosetEventKind
+  sealed trait EndKind extends InfosetEventKind
+  sealed trait ArrayKind extends InfosetEventKind
+  sealed trait ElementKind extends InfosetEventKind
+  case object StartElement extends StartKind with ElementKind
+  case object EndElement extends EndKind with ElementKind
+  case object StartArray extends StartKind with ArrayKind
+  case object EndArray extends EndKind with ArrayKind
+}
 
 /**
  * An infoset event accessor.
  *
  * Not a case class because we don't want pattern matching on this. (It allocates to pattern match)
  *
- * We have erd as a seperate field (instead of just using node.erd) since
- * the inspectPure method finds an erd without constructing the corresponding DINode (so as to avoid side effects)
+ * The info associated with an infoset event is either a DIElement (for elements), or
+ * an ElementRuntimeData for arrays. We don't want to use a DIArray object because those
+ * are the private managed storage of DIComplex elements.
+ *
+ * At the time these info DIElements are created to carry the infoset event information
+ * they are not attached to parents/children in the infoset.
+ *
+ * We are effectively just using them as event "tuples" to carry the ERD and simple value
+ * information after having resolved the incoming event's local name + namespace (if we're using those)
+ * to get the ERD.
+ *
+ * By using our DIElement objects as the "tuples" of event information, we avoid having to
+ * allocate yet another separate object for these.
  *
- * We maintaint the invariant that node==null || node.erd==erd
+ * The Unparsers will examine the incoming DIElement, and associate them parent with child if
+ * they decide to keep them.
  *
+ * Unparsers are also responsible for looking at the incoming event, and deciding if we
+ * should first fill in any defaultable elements, or compute any dfdl:outputValueCalc elements.
+ *
+ * This changes the behavior of infoset inputters. You can no longer just pull events and have them
+ * get connected together into an infoset.
+ *
+ * Note that populating an accessor does NOT perform any side-effects on the infoset.
+ * That is left up to the Unparser combinators, largely the Element Combinators.
  */
```

Comment:
This is making me wonder if the infoset event stuff is even really needed anymore? THe unparser a tied much more closely to the infoset inputter now. For example, they are now triggering the creation of new elements, whereas previously this was all handled by the infoset inputter and the unparsers could just assume the infoset nodes were created automatically.

I haven't put much thought into this, so this might be completely wrong, but is there any benefit to removing the infoset event stuff, and just moving more of this logic into  individual unparsers? I feel like the infoset accessor stuff was to support streaming and decouple unparsers from infoset creation, but that seems to not be the case anymore.

So I guess the question is, does it make sense to have Unparsers directly access the InfosetInputter without this indirection from the accessor?

Reply:
The infoset inputter still feels like a useful abstraction to me. It still creates the infoset elements and resolves the elements. It also solves the problem of inferring array boundaries (start/end) from occurrences of element start/end, which is necessary for XML.  

One point of criticism is that it doesn't make it simpler if a representation has its own explicit arrays, as JSON does. In some sense, InfosetInputters should have startArray() and endArray() methods that are directly called for JSON, and for XML, the inputter synthesizes and calls these event callbacks when it infers the array boundaries.

Instead, we treat all infoset event sources as if they were "as dumb as" XML, and always infer array boundaries. This is ok, as our infoset is XML-like. E.g., has no notion of Array with no elements in it. JSON has empty arrays as first class things, but we can't represent that. 

The accessor is just an internal API thing to avoid allocation of little event tuple objects all the time. You advance/inspect and you get back two things: start/end, and "info" which is arrayERD, or a DIElement.